yogstation/code/modules/mob/say.dm

```diff
@@ -1,3 +1,73 @@
+#define TYPING_INDICATOR_RANGE 7
+
 /mob/proc/get_say()
+	create_typing_indicator()
+	window_typing = TRUE
 	var/msg = input(src, null, "say \"text\"") as text|null
-	say_verb(msg)
\ No newline at end of file
+	window_typing = FALSE
+	remove_typing_indicator()
+	say_verb(msg)
+
+/mob
+	var/image/typing_overlay
+	var/list/speech_bubble_recipients
+	var/last_typed
+	var/last_typed_time
+	var/window_typing = FALSE
+	var/bar_typing = FALSE
+
+/mob/proc/handle_typing_indicator()
+	if(!GLOB.typing_indicators)
+		return
+	if(client)
+		var/temp = winget(client, "input", "text")
+		if(temp != last_typed)
+			last_typed = temp
+			last_typed_time = world.time
+		if(world.time > last_typed_time + 10 SECONDS)
+			bar_typing = FALSE
+			remove_typing_indicator()
+			return
+		if(length(temp) > 5 && findtext(temp, "Say \"", 1, 7))
+			create_typing_indicator()
+			bar_typing = TRUE
+		else if(length(temp) > 3 && findtext(temp, "Me ", 1, 5))
+			//set_typing_indicator(1)
+		else
+			bar_typing = FALSE
+			remove_typing_indicator()
+			
+
+/mob/proc/create_typing_indicator()
```

Comment:
I would recommend a blanking overwrite of this for ``/mob/camera``, since otherwise things like the AI camera eye and the Eminence have a visible typing indicator, despite them needing to be invisible:
```dm
/mob/camera/create_typing_indicator()
  return
/mob/camera/remove_typing_indicator()
  return
```

Reply:
> I would recommend a blanking overwrite of this for `/mob/camera`, since otherwise things like the AI camera eye and the Eminence have a visible typing indicator, despite them needing to be invisible:
> 
> ```dm
> /mob/camera/create_typing_indicator()
>   return
> /mob/camera/remove_typing_indicator()
>   return
> ```

This is solved by the overlays inheriting the invisibility of the mob they are attached to in regards to the wrong people seeing them. 

On one hand I like the idea that the people who can see them can also see if they're typing